cpp/oneapi/dal/backend/primitives/super_accumulator/int128.hpp

```diff
@@ -0,0 +1,224 @@
+
+union duality64 {
+    double floatingpoint;
+    std::uint64_t integer;
+    constexpr duality64(const std::uint64_t& val) : integer(val) {}
+    constexpr duality64(const double& val) : floatingpoint(val) {}
+};
+
+template <typename Impl>
+class int128_base {
+public:
+    using derived_t = Impl;
+
+    int128_base() = default;
+
+    std::int64_t& get_high_half() {
+        return static_cast<Impl*>(this)->get_high_half_impl();
+    }
+
+    std::uint64_t& get_low_half() {
+        return static_cast<Impl*>(this)->get_low_half_impl();
+    }
+
+    const std::int64_t& get_high_half() const {
+        return static_cast<const Impl*>(this)->get_high_half_impl();
+    }
+
+    const std::uint64_t& get_low_half() const {
+        return static_cast<const Impl*>(this)->get_low_half_impl();
+    }
+
+    static derived_t make(std::int64_t h, std::uint64_t l, derived_t result = {}) {
+        result.get_high_half() = h;
+        result.get_low_half() = l;
+        return result;
+    }
+
+    template <typename AnotherImpl, bool atomic = false>
+    Impl& add(const int128_base<AnotherImpl>& rhs) {
+        if constexpr (atomic) {
+#ifdef __SYCL_DEVICE_ONLY__
+            const std::uint64_t low = atomic_global_add(&get_low_half(), rhs.get_low_half());
+            // We need to add number again because atomic_fetch_add returns prev value
+            const bool carry = bool((low + rhs.get_low_half()) < rhs.get_low_half());
+            const auto incrt = rhs.get_high_half() + carry;
+            atomic_global_add(&get_high_half(), incrt);
+#else
+            const std::uint64_t low = (get_low_half() += rhs.get_low_half());
+            const bool carry = bool(low < rhs.get_low_half());
+            get_high_half() += rhs.get_high_half() + carry;
+#endif
+        }
+        else {
+            get_low_half() += rhs.get_low_half();
+            bool carry = bool(get_low_half() < rhs.get_low_half());
+            get_high_half() += rhs.get_high_half() + carry;
+        }
+        return *static_cast<Impl*>(this);
+    }
+
+    operator double() const {
+        return to_floatingpoint<double>();
+    }
+
+protected:
+    template <typename Float>
+    auto to_floatingpoint() const {
+        return (get_high_half() < 0l) ? -to_floatingpoint_pos<Float>(-(*this))
+                                      : to_floatingpoint_pos<Float>(*this);
+    }
+
+    template <typename Float, typename AnotherImpl>
+    static auto to_floatingpoint_pos(const int128_base<AnotherImpl>& x) {
+        constexpr duality64 mul{ static_cast<std::uint64_t>(0x43f0000000000000ul) };
+        return static_cast<Float>(x.get_low_half()) +
+               static_cast<Float>(x.get_high_half()) * mul.floatingpoint;
+    }
+
+private:
+#ifdef ONEDAL_DATA_PARALLEL
+    template <typename T>
+    static inline T atomic_global_add(T* ptr, T operand) {
+        using address = cl::sycl::access::address_space;
+        return cl::sycl::atomic_fetch_add<T, address::global_space>(
```

Comment:
Deprecated in SYCL 2020. Use atomic_ref instead

Reply:
Okay. I'll move to refs later